Add surface_t and use it in graphics.c and rdp.c

rasky opened the pull request:
Another partial merge coming from the rdpq/opengl branch.

This introduces the new `surface_t` type and use it everywhere. Thanks to a trick, also display.c
will start using it in a backward compatible fashion.

See the descriptions of commit by commit for further information.

rasky committed:
Add surface_t structure to represent a generic graphic surface.

This will come useful as we improve RDP support in libdragon. Currently
we have a "display_context_t" which is just a framebuffer handled by
display.c, and a "sprite_t" which is an owned bitmap. 

An optionally-owned surface, allowing also for sub-surfaces, is a more
generic paradigm that can make for more flexible implementations.

rasky committed:
Add color_t function helpers

There is currently no helpers to manage a color_t structure. As we are
going to make large use of it in the new RDP library, start by providing
a few helpers.

rasky committed:
display: support more than 3 buffers.

Currently, display.c is limited to either 2 or 3 framebuffers. This
commit lifts the upper limit, by handling a bitmask of locked buffers
and a bitmask of "ready to show" buffers.

While for a game it is probably not wise to increase this too much
(as input latency would grow), there are other use cases where it can
become very useful, like full motion videos that can now pre-buffer
more frames.

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+ * 
+ * A surface is described by the following properties:
+ * 
+ *  * Size (width. height)
```
-> "width, height"

rasky replied:
done

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+ *      graphics_draw_text(&buf, 0, 0, "ABC");
+ * @endcode
+ *
+ * Sometimes, you might have an existing raw pointer to a buffer and need to pass it
+ * to an API that accepts a #surface_t. For those cases, you can use
+ * #surface_make to create a #surface_t instance, that you can throw away
+ * after you called the function.
```
It's probably worth mentioning that you can safely call `surface_free` on this kind of surface as well (Same with the sub-surface below)

rasky replied:
done

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+ * Note that there are texture format that are 4bpp, so don't divide this by 8 to get the number of bytes
+ * per pixels, but rather use #TEX_FORMAT_BYTES2PIX and #TEX_FORMAT_PIX2BYTES. */
+#define TEX_FORMAT_BITDEPTH(fmt)             (4 << ((fmt) & 0x3))
+/** @brief Convert the specified number of pixels in bytes. */
```
-> "to bytes"

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+#define TEX_FORMAT_BITDEPTH(fmt)             (4 << ((fmt) & 0x3))
+/** @brief Convert the specified number of pixels in bytes. */
+#define TEX_FORMAT_PIX2BYTES(fmt, pixels)    ((pixels) << (((fmt) & 3) + 2) >> 3)
+/** @brief Convert the specified number of bytes in pixels. */
```
-> "to pixels"

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+ * Note that there are texture format that are 4bpp, so don't divide this by 8 to get the number of bytes
+ * per pixels, but rather use #TEX_FORMAT_BYTES2PIX and #TEX_FORMAT_PIX2BYTES. */
+#define TEX_FORMAT_BITDEPTH(fmt)             (4 << ((fmt) & 0x3))
+/** @brief Convert the specified number of pixels in bytes. */
+#define TEX_FORMAT_PIX2BYTES(fmt, pixels)    ((pixels) << (((fmt) & 3) + 2) >> 3)
+/** @brief Convert the specified number of bytes in pixels. */
+#define TEX_FORMAT_BYTES2PIX(fmt, bytes)     (((bytes) << 1) >> ((fmt) & 3))
```
We should probably mention that both of these macros always round down. Also I think that a version of `TEX_FORMAT_PIX2BYTES` that rounds up is needed (or maybe it should do that by default). For example one might want to allocate a 4 bit per pixel surface with an odd width, in which case it should round up to the next byte. In fact, that should probably be done in `surface_alloc`.

rasky replied:
I think it makes sense to always roundup `TEX_FORMAT_PIX2BYTES`. For `TEX_FORMAT_BYTES2PIX`, the current rounding-down behavior should symmetrically be the correct one.

I'ved added the roundup and updated the documentation.

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+ * @see #surface_make
+ */
+inline surface_t surface_make_linear(void *buffer, tex_format_t format, uint32_t width, uint32_t height) {
+    return surface_make(buffer, format, width, height, TEX_FORMAT_PIX2BYTES(format, width));
```
What to do here if the format is 4 bits per pixel and the width is odd?

rasky replied:
In the latest version, the stride is rounded up, which I think makes sense in the case as well.

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+ * 
+ * A surface allocated via #surface_alloc can be used as a RDP frame buffer
+ * (passed to #rdp_attach) because it is guarateed to have the required
+ * alignment of 64 bytes.
```
Provided it uses one of the supported formats.

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+ *        surface.
+ * 
+ * The surface returned by this function will point to a portion of the buffer of
+ * the parent surface, and will have of course the smae pixel format.
```
-> "same"

snacchus commented on include/surface.h:
```diff
@@ -0,0 +1,245 @@
+ * 
+ * @param[in]  parent   Parent surface that will be pointed to
+ * @param[in]  x0       X coordinate of the top-left corner of the parent surface
+ * @param[in]  y0       Y coordinate of the top-left corner of the parent surface
```
Technically, they are the coordinates of the new sub-surface within the parent surface. The wording might need to be clarified.

snacchus commented on src/display.c:
```diff
@@ -162,27 +166,32 @@ static const uint32_t * const reg_values[] = {
 
-/** @brief Complete drawn buffer to display next */
-static int show_next = -1;
+/** @brief True if the buffer indexed by now_drawing is currently locked */
```
Is this doc still correct?

rasky replied:
good catch, fixed

snacchus commented on src/display.c:
```diff
@@ -479,42 +486,53 @@ void display_close()
+        if( __safe_buffer[i] )
         {
-            free( buffer[i]);
+            free_uncached( __safe_buffer[i]);
```
Why not use `surface_free` instead?

snacchus commented on src/display.c:
```diff
@@ -479,42 +486,53 @@ void display_close()
+display_context_t display_lock(void)
 {
-    display_context_t retval = 0;
+    display_context_t retval = NULL;
```
Update to use `surface_t *` as well

snacchus commented on include/display.h:
```diff
@@ -65,18 +70,27 @@ typedef enum
+ * @brief Display context (DEPRECATED: Use #surface_t instead)
+ * 
+ * @see #surface_t
+ */
+typedef surface_t* display_context_t;
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 void display_init( resolution_t res, bitdepth_t bit, uint32_t num_buffers, gamma_t gamma, antialias_t aa );
-display_context_t display_lock();
-void display_show(display_context_t disp);
+surface_t* display_lock(void);
+void display_show(surface_t* disp);
 void display_close();
 
+uint32_t display_get_width(void);
+uint32_t display_get_height(void);
+uint32_t display_get_bitdepth(void);
+uint32_t display_get_num_buffers(void);
```
This is probably a good opportunity to move the docs to the header file.

rasky replied:
done

snacchus replied:
Have you committed this yet? It's still in the .c file for me.

rasky replied:
I reverted it and forgot to mention it. I will do this in a final commit before merging otherwise there are too many conflicts in case other changes are needed

rasky replied:
Added now

snacchus commented on src/surface.c:
```diff
@@ -0,0 +1,61 @@
+    tex_format_t fmt = surface_get_format(parent);
+
+    surface_t sub;
+    sub.buffer = parent->buffer + y0 * parent->stride + TEX_FORMAT_PIX2BYTES(fmt, x0);
```
Again, what to do if width is odd and the surface has a 4 bits per pixel format?

rasky replied:
It's not possible to make a sub surface starting from an odd offset in a 4bpp format, for obvious reasons. For even offsets, this should work now.

snacchus replied:
So if `x0` is odd, then this will silently set the beginning of the buffer to the next byte. We should probably assert in that case, because this could have unforeseen consequences (like overflowing at the end of the line). And obviously the first pixel of each line is not what the user might expect.

rasky replied:
Done, see if it seems OK to you.

snacchus replied:
Looks good.

snacchus commented on src/surface.c:
```diff
@@ -0,0 +1,61 @@
+        .flags = format | SURFACE_FLAGS_OWNEDBUFFER,
+        .width = width,
+        .height = height,
+        .stride = TEX_FORMAT_PIX2BYTES(format, width),
+        .buffer = malloc_uncached_aligned(64, height * TEX_FORMAT_PIX2BYTES(format, width)),
```
See my comment in surface.h

rasky replied:
This should be fixed as well with the roundup.

snacchus commented on src/display.c:
```diff
@@ -479,42 +486,53 @@ void display_close()
 /**
  * @brief Lock a display buffer for rendering
  *
  * Grab a display context that is safe for drawing.  If none is available
```
Should we still use the term "display context" in documentation or just switch to "surface"?

snacchus reviewed (commented):
Implementation looks good to me, my only concern is with `TEX_FORMAT_PIX2BYTES` in combination with 4 bit per pixel formats (see comments). Otherwise made some comments on documentation.

rasky committed:
display: change display_context_t into a surface_t*

This is a "hopefully backward compatible change". Before this commit,
display.c returned display_context_t which was an opaque index into an
internal array of display buffers, or 0 to indicate "no buffer".

Change the typedef so that the type is no more an integer but a
surface_t pointer. This should work transparently for all users that
have correctly used display_context_t to type the return type of
display_lock. Users that have manually written "int" will get a
compilation error, but with an easy fix: just change the type. Notice
that "0" indicating an error still works for a pointer as well.

The commit also adapts graphics.c and rdp.c to accept directly
a surface_t pointer, making them useful in different context: in fact,
now they can be used not only to draw into the display.c framebuffers,
but into any surface like offscreen render targets.

Rename also `rdp_attach_display` to `rdp_attach` to reflect the
increased scope of the function.

rasky commented:
@snacchus thanks for the review, please double check the odd/even stuff with 4bpp

rasky committed:
display: move docs of public API from c to h
